### PR TITLE
feat(ci): add cosign to sign the released binaries

### PR DESCRIPTION
**Pull request by Mossaka**
This draft PR adds cosign signing for the released binaries. It is WIP.

I am testing releases in my own forked repo: https://github.com/Mossaka/runwasi/actions/runs/7634841458

Related to #417

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -38,30 +41,70 @@ jobs:
...
         if: ${{ matrix.arch == 'x86_64' }}
         timeout-minutes: 10
         run: make test-${{ needs.parse.outputs.runtime }}
+
+      - name: Sign the binary
+        if: ${{ needs.parse.outputs.runtime != 'wasm' }}
+        run: |
+          make dist-${{ needs.parse.outputs.runtime }}
+          # Check if there's any files to archive as tar fails otherwise
+          if stat dist/bin/* >/dev/null 2>&1; then
```
Comment (jsturtevant) [inline .github/workflows/release.yml:81]: What's the scenario where there are no files after running make `dist-*`

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -29,7 +29,10 @@ jobs:
         shell: bash
         run: ./scripts/parse_ref.sh ${{ github.ref }} >> ${GITHUB_OUTPUT}
 
-  build:
+  build-and-sign:
+    permissions:
+      # cosign uses the GitHub OIDC token
+      id-token: write
```
Comment (jsturtevant) [inline .github/workflows/release.yml:35]: What account is this set up for?

**Comment by Mossaka** [comment]
> Could we add a doc on how to run the verification after downloading the files?

Sure, I also intend to add the verification doc on the release page.

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -38,30 +41,70 @@ jobs:
...
+              --output-certificate containerd-${{ needs.parse.outputs.runtime }}d-crt.pem \
+              --output-signature containerd-${{ needs.parse.outputs.runtime }}d.sig \
+              dist/bin/containerd-${{ needs.parse.outputs.runtime }}d
+          else
+            echo "No files to sign"
+          fi
+
+          # Copy the certs to the dist/bin folder
+          cp *.sig dist/bin/
+          cp *.pem dist/bin/
```
[the remark this answers is not in the document]
Comment (Mossaka) [reply .github/workflows/release.yml:102]: I will move them into the `if` statement

File: .github/workflows/release.yml (hunk 1 above)
Comment (Mossaka) [reply .github/workflows/release.yml:81]: You meant afer `make build-*`? 

If we release `containerd-shim-wasm-test-modules`, I think there will be no files in `disk/bin`.

File: .github/workflows/release.yml (hunk 2 above)
Comment (Mossaka) [reply .github/workflows/release.yml:35]: The `id-token: write` permission enables the gh workflow to request an OIDC token generated from GitHub OIDC provider, which will then be used by `cosign` to authenticate with Sigstore for signing. `cosign` uses Fulcio, upon receiving the OIDC token, to validate the token. This validation includes checking the token's integrity, its issuer (GitHub in this case), and the claims within the token (like the repository and job details). 

After the validation step, Fulcio trusts that the signing request is legitimately coming from the specified GitHub Actions workflow.

So to answer your question, it does not set up an account for anyone. `cosign` backend verfies the emphameral token with GitHub OIDC provider.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -38,30 +41,70 @@ jobs:
...
+          # Check if there's any files to archive as tar fails otherwise
+          if stat dist/bin/* >/dev/null 2>&1; then
+            cosign sign-blob --yes \
+              --output-certificate containerd-shim-${{ needs.parse.outputs.runtime }}-v1-crt.pem \
+              --output-signature containerd-shim-${{ needs.parse.outputs.runtime }}-v1.sig \
+              dist/bin/containerd-shim-${{ needs.parse.outputs.runtime }}-v1
+            
+            cosign sign-blob --yes \
+              --output-certificate containerd-shim-${{ needs.parse.outputs.runtime }}d-v1-crt.pem \
+              --output-signature containerd-shim-${{ needs.parse.outputs.runtime }}d-v1.sig \
```
[the remark this answers is not in the document]
Comment (Mossaka) [reply .github/workflows/release.yml:89]: Ah I guess the bundle format is simplier in this case. Thanks

File: .github/workflows/release.yml (hunk 2 above)
Comment (jsturtevant) [reply .github/workflows/release.yml:35]: the verification doc might help clarify it for me

Comment (Mossaka) [reply .github/workflows/release.yml:35]:
```
cosign verify-blob --signature containerd-shim-wasmtime-v1.sig --certificate containerd-shim-wasmtime-v1-crt.pem  --certificate-identity https://github.com/Mossaka/runwasi/.github/workflows/release.yml@refs/tags/containerd-shim-wasmtime/v0.3.1 --certificate-oidc-issuer https://token.actions.githubusercontent.com  containerd-shim-wasmtime-v1
```

This is the command I used to verify the release binary on a forked repo when I was testing this PR.

Comment (Mossaka) [reply .github/workflows/release.yml:35]: As the command shows, the identity is a reference to that workflow run which requests the token from Github OIDC. 
`--certificate-identity https://github.com/Mossaka/runwasi/.github/workflows/release.yml@refs/tags/containerd-shim-wasmtime/v0.3.1`

Comment (Mossaka) [reply .github/workflows/release.yml:35]: Added instructions to verify blobs on RELEASE.MD

Comment (jsturtevant) [reply .github/workflows/release.yml:35]: > As the command shows, the identity is a reference to that workflow run which requests the token from Github OIDC. `--certificate-identity https://github.com/Mossaka/runwasi/.github/workflows/release.yml@refs/tags/containerd-shim-wasmtime/v0.3.1`

got it thanks!  This does seem a little strange to me to be the cert identity 🤷

**Comment by Mossaka** [comment]
@jeremyrickard could you please take a look?

**Review by jsturtevant — COMMENTED** [review]
LGTM, it would be nice to have one more set of eyes on this PR since I am fairly new to cosign

**Review by jeremyrickard — APPROVED** [review]
LGTM

**Pull request merged**